### PR TITLE
Add frontend unit testing (Vitest + @testing-library)

### LINKED ISSUE #432: Add frontend unit testing (Vitest + @testing-library)
Part of #431.

## Problem

The frontend has **no unit test layer** — only Playwright E2E (`frontend/e2e/`). The starter-kit prescribes Vitest for unit tests. Heavy/critical frontend logic is currently covered only end-to-end (or not at all):

- graphology layout / d3 / elk graph code in `src/components/network/`
- data hooks (`src/features/network/hooks/useNetworkData.ts`, the recently-hardened entity hooks)
- `src/services/api/client.ts`

## Proposal

- Add `vitest`, `@testing-library/react`, `@testing-library/jest-dom`, `jsdom` as devDependencies.
- Wire Vitest into the existing `vite.config.ts` (`test` block, jsdom environment) — no separate config needed.
- Add `test` and `test:watch` scripts to `frontend/package.json`.
- Add one representative test (e.g. an entity hook or a pure util) to prove the harness.
- Add a CI job (mirror `.github/workflows/test-backend.yml`) running `npm run test` on the frontend.

## Acceptance criteria

- [ ] `npm run test` runs Vitest green locally
- [ ] At least one meaningful unit test committed
- [ ] CI job runs frontend unit tests on PRs
- [ ] `docker compose up -d --build` still succeeds

### PR DESCRIPTION
**Pull request by NotYuSheng**
## Summary
- Add `vitest`, `@testing-library/react`, `@testing-library/jest-dom`, and `jsdom` as devDependencies
- Wire Vitest into `vite.config.ts` (jsdom environment, globals, e2e exclusion) with `test` and `test:watch` scripts
- Add 52 unit tests across 4 utility modules (`formatters`, `portUtils`, `dateUtils`, `deviceType`) to prove the harness
- Add CI workflow (`.github/workflows/test-frontend.yml`) mirroring the backend test job

Closes #432

## Test plan
- [x] `npm run test` runs Vitest green locally (52 tests passing)
- [x] `npx tsc --noEmit` passes from `frontend/`
- [x] `docker compose up -d --build` succeeds
- [ ] CI workflow runs on this PR and passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit test suite for utility functions covering date parsing, device type handling, data formatting, and port utilities.
  * Configured automated frontend testing with GitHub Actions workflow that runs on pull requests.

### CODE REVIEW

File: frontend/vite.config.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/// <reference types="vitest/config" />
 import { defineConfig, loadEnv } from 'vite'
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/vite.config.ts:2]: Good catch — imported `configDefaults` from `vitest/config` and spread it into the `exclude` array so default exclusions (node_modules, dist, .git, etc.) are preserved. Fixed in 883b9ad.

File: frontend/vite.config.ts [hunk 2]
```diff
@@ -55,6 +56,13 @@ export default defineConfig(({ mode }) => {
         },
       },
     },
+    test: {
+      globals: true,
+      environment: 'jsdom',
+      setupFiles: './src/test/setup.ts',
+      css: true,
+      exclude: ['e2e/**', 'node_modules/**'],
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/vite.config.ts]: Agreed — now using `[...configDefaults.exclude, 'e2e/**']` to extend rather than replace the defaults. Fixed in 883b9ad.

File: frontend/src/utils/__tests__/dateUtils.test.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { parseDateTime } from '../dateUtils';
+
+describe('parseDateTime', () => {
+  it('parses an ISO date string', () => {
+    const ts = parseDateTime('2025-06-15T10:30:00');
+    expect(ts).toBe(new Date('2025-06-15T10:30:00').getTime());
+  });
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts]: Valid point about the tautology. Updated to use a UTC ISO string (`2025-06-15T10:30:00Z`) with a hardcoded epoch timestamp. Note: the correct value is `1749983400000`, not `1750012200000` as suggested. Fixed in 883b9ad.

File: frontend/src/utils/__tests__/dateUtils.test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+describe('parseDateTime', () => {
+  it('parses an ISO date string in UTC', () => {
+    const ts = parseDateTime('2025-06-15T10:30:00Z');
+    expect(ts).toBe(1749983400000);
+  });
+
+  it('parses a Spring LocalDateTime array [Y, M, D, h, m, s]', () => {
+    const ts = parseDateTime([2025, 6, 15, 10, 30, 0]);
+    expect(ts).toBe(new Date(2025, 5, 15, 10, 30, 0).getTime());
+  });
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts:18]: Agreed the test was a tautology. Replaced with component-level assertions (year, month, date, hours, minutes, seconds) that verify the month-offset mapping and field ordering without being timezone-dependent.

Regarding the suggestion to change `parseDateTime` itself to use `Date.UTC`: Spring's `LocalDateTime` arrays are intentionally interpreted as local time — the server and client share the same timezone (Asia/Singapore). Changing to `Date.UTC` would shift all displayed times by 8 hours and is a behavioral change beyond the scope of this PR. Fixed test in 9c27244.

File: frontend/vite.config.ts [hunk 5]
```diff
@@ -55,6 +57,13 @@ export default defineConfig(({ mode }) => {
         },
       },
     },
+    test: {
+      globals: true,
+      environment: 'jsdom',
+      setupFiles: './src/test/setup.ts',
+      css: true,
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/vite.config.ts]: Good call — the current suite is all pure utility tests and doesn't need CSS parsing. Switched to `css: false`. Fixed in 9c27244.

File: frontend/src/utils/__tests__/dateUtils.test.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    expect(date.getDate()).toBe(15);
+    expect(date.getHours()).toBe(10);
+    expect(date.getMinutes()).toBe(30);
+    expect(date.getSeconds()).toBe(0);
+  });
+
+  it('returns 0 for a short array', () => {
+    expect(parseDateTime([2025, 6])).toBe(0);
+  });
+});
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts:38]: Added — `parseDateTime('invalid-date')` now asserts `toBeNaN()`. Fixed in 7f7b35f.

File: frontend/src/utils/__tests__/formatters.test.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  it('formats minutes and seconds', () => {
+    expect(formatDuration(150000)).toBe('2m 30s');
+  });
+
+  it('formats hours and minutes', () => {
+    expect(formatDuration(3660000)).toBe('1h 1m');
+  });
+});
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/formatters.test.ts:55]: Added boundary cases for 0ms, exactly 1 minute, and exactly 1 hour. Fixed in 7f7b35f.

File: frontend/vite.config.ts [hunk 8]
```diff
@@ -55,6 +57,13 @@ export default defineConfig(({ mode }) => {
         },
       },
     },
+    test: {
+      globals: true,
+      environment: 'jsdom',
+      setupFiles: './src/test/setup.ts',
+      css: false,
+      exclude: [...configDefaults.exclude, 'e2e/**'],
+    },
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/vite.config.ts:66]: Acknowledged — this is a reasonable perf optimization. However, keeping `jsdom` as default ensures future component tests work without extra boilerplate. With 54 pure-utility tests the full suite runs in <1s, so the overhead is negligible for now. Worth revisiting once the component test count grows significantly.

File: .github/workflows/test-frontend.yml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    name: vitest run
+    runs-on: ubuntu-latest
+
+    defaults:
+      run:
+        working-directory: frontend
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply .github/workflows/test-frontend.yml:24]: Valid security hardening, but the existing `test-backend.yml` also uses `actions/checkout@v4` without `persist-credentials: false`. Applying this only to the frontend CI would be inconsistent — this should be done as a repo-wide hardening pass across all workflows.

File: .github/workflows/test-frontend.yml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+      - name: Set up Node
+        uses: actions/setup-node@v4
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply .github/workflows/test-frontend.yml:27]: Acknowledged — SHA pinning is a best practice. However, `test-backend.yml` also uses `@v4` tags for `actions/checkout` and `actions/setup-java`. Pinning only the frontend workflow would be inconsistent. This should be addressed as a separate repo-wide CI hardening task.

File: frontend/src/utils/__tests__/dateUtils.test.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  it('parses a Spring LocalDateTime array [Y, M, D, h, m, s]', () => {
+    const ts = parseDateTime([2025, 6, 15, 10, 30, 0]);
+    const date = new Date(ts);
+    expect(date.getFullYear()).toBe(2025);
+    expect(date.getMonth()).toBe(5);
+    expect(date.getDate()).toBe(15);
+    expect(date.getHours()).toBe(10);
+    expect(date.getMinutes()).toBe(30);
+    expect(date.getSeconds()).toBe(0);
+  });
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts:18]: This was already addressed in round 2 — Spring's `LocalDateTime` arrays are intentionally interpreted as local time because the server and client share the same timezone (Asia/Singapore, set in Docker config). The ISO string path uses UTC which is correct for absolute timestamps. Changing `parseDateTime` to use `Date.UTC` for arrays would shift all displayed times by 8 hours. This is a deliberate design choice, not a bug.

File: frontend/src/utils/__tests__/dateUtils.test.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+  it('returns 0 for a short array', () => {
+    expect(parseDateTime([2025, 6])).toBe(0);
+  });
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts:33]: Valid observation about 5-element arrays from Spring's Jackson serialization. However, this is a functional enhancement to `parseDateTime` itself, not a test issue — beyond the scope of this PR which adds test infrastructure. Worth a follow-up issue.

File: frontend/src/utils/__tests__/formatters.test.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+describe('formatIpPort', () => {
+  it('returns IP with port when port is provided', () => {
+    expect(formatIpPort('192.168.1.1', 80)).toBe('192.168.1.1:80');
+  });
+
+  it('returns IP only when port is omitted', () => {
+    expect(formatIpPort('10.0.0.1')).toBe('10.0.0.1');
+  });
+});
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/formatters.test.ts:90]: Valid point about IPv6 bracket notation. This is a functional enhancement to `formatIpPort` itself — out of scope for the test scaffolding PR. Worth a follow-up.

File: frontend/src/utils/__tests__/dateUtils.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/dateUtils.test.ts:33]: Acknowledged — the inconsistency between returning `0` for short arrays vs `NaN` for invalid strings is a pre-existing design quirk in `parseDateTime`. Normalizing error returns across all invalid inputs is a worthwhile follow-up but beyond the scope of this test-scaffolding PR.

File: frontend/src/utils/__tests__/formatters.test.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  it('formats hours and minutes', () => {
+    expect(formatDuration(3660000)).toBe('1h 1m');
+  });
+
+  it('handles boundary values', () => {
+    expect(formatDuration(0)).toBe('0ms');
+    expect(formatDuration(60000)).toBe('1m 0s');
+    expect(formatDuration(3600000)).toBe('1h 0m');
+  });
+});
```
[the remark this answers is not in the document]
Comment (NotYuSheng) [reply frontend/src/utils/__tests__/formatters.test.ts:55]: Acknowledged — negative inputs are a pre-existing edge case in these formatters. In practice they're never called with negatives (bytes and durations are always non-negative), but hardening them is a reasonable follow-up.
